django/views/debug.py

```diff
@@ -459,14 +468,7 @@ def explicit_or_implicit_cause(exc_value):
                 'post_context': post_context,
                 'pre_context_lineno': pre_context_lineno + 1,
             })
-
-            # If the traceback for current exception is consumed, try the
-            # other exception.
-            if not tb.tb_next and exceptions:
-                exc_value = exceptions.pop()
-                tb = exc_value.__traceback__
-            else:
-                tb = tb.tb_next
+            tb = tb.tb_next
```

Comment:
What do you think about moving the logic from the loop to a separate hook (as suggested by Chris)?, this will create a separation between iteration through exceptions and their tracebacks e.g.
```python

    def get_traceback_frames(self):
        ...

        frames = []
        # No exceptions were supplied to ExceptionReporter
        if not exceptions:
            return frames

        # In case there's just one exception, take the traceback from self.tb
        exc_value = exceptions.pop()
        tb = self.tb if not exceptions else exc_value.__traceback__
        frames.extend(self.get_inner_traceback_frames(exc_value, tb))
        while exceptions:
            exc_value = exceptions.pop()
            exc_frames = self.get_inner_traceback_frames(exc_value, exc_value.__traceback__)
            frames.extend(exc_frames)
        return frames

    def get_inner_traceback_frames(self, exc_value, tb):
        frames = []
        exc_cause = self.explicit_or_implicit_cause(exc_value),
        exc_cause_explicit = getattr(exc_value, '__cause__', True)
        
        while tb is not None:
            # Support for __traceback_hide__ which is used by a few libraries
            ...
            frames.append({
                'exc_cause': exc_cause,
                'exc_cause_explicit': exc_cause_explicit,
                'tb': tb,            
                'type': 'django' if module_name.startswith('django.') else 'user',
                'filename': filename,
                'function': function,
                'lineno': lineno + 1,
                'vars': self.filter.get_traceback_frame_variables(self.request, tb.tb_frame),
                'id': id(tb),        
                'pre_context': pre_context,
                'context_line': context_line,
                'post_context': post_context,
                'pre_context_lineno': pre_context_lineno + 1,
            })
            tb = tb.tb_next          
        return frames
```

Probably, this could be simplified even more.

Reply:
I updated the code to have just one loop over the exceptions.
By this change, the `view_tests.tests.test_debug.ExceptionReporterTests.test_reporting_frames_for_cyclic_reference
` will fail. I can see the `ExceptionCycleWarning` in console but I don't know what's the problem :thinking:

Reply:
I didn't suggest using a single loop, but two nested loops, first by exceptions and the second by traceback frames for each exception. Using a single loop is not readable and causes tests failures.

Reply:
Thanks @felixxm. I changed it to two nested loops.